### PR TITLE
Public holiday controller service route function

### PR DESCRIPTION
**Pull request by clinton-otse**

### CODE REVIEW

File: src/services/public-holiday/public-holiday.service.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  public async create(req, publicHolidayData: CreatePublicHolidayDto): Promise<IPublicHoliday> {
+    const newData = publicHolidayData;
+
+    const publicHolidayExists = await publicHolidayModel.exists(newData);
+    if (publicHolidayExists) {
+      throw new HttpException(409, `${newData.title} Public holiday already exists`);
+    }
+
+    newData.created_by = req.user._id;
+    newData.project_id = req.user.projectId._id;
```
Comment (anthugny) [inline src/services/public-holiday/public-holiday.service.ts:20]: the project id comes from the Public Holiday DTO.

File: src/services/public-holiday/public-holiday.service.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import omit from 'lodash/omit';
+import { isEmpty } from '@/utils/util';
+
+class PublicHolidayService {
+  public publicHolidayModel = publicHolidayModel;
+
+  public async create(req, publicHolidayData: CreatePublicHolidayDto): Promise<IPublicHoliday> {
+    const newData = publicHolidayData;
+
+    const publicHolidayExists = await publicHolidayModel.exists(newData);
```
Comment (anthugny) [inline src/services/public-holiday/public-holiday.service.ts:14]: check with newData.title. You can make all titles lower case before saving.

File: src/services/public-holiday/public-holiday.service.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    if (publicHolidayExists) {
+      throw new HttpException(409, `${newData.title} Public holiday already exists`);
+    }
+
+    newData.created_by = req.user._id;
+    newData.project_id = req.user.projectId._id;
+
+    let publicHoliday = await publicHolidayModel.create(newData);
+    console.log(`${newData.title} Public holiday has been created`, publicHoliday);
```
Comment (anthugny) [inline src/services/public-holiday/public-holiday.service.ts:23]: clear out your console.logs before pushing.

File: src/services/public-holiday/public-holiday.service.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    if (!publicHoliday) {
+      throw new HttpException(404, `Public holiday not found`);
+    }
+    return publicHoliday;
+  }
+
+  public async update(id: string, updatedData: IPublicHoliday): Promise<IPublicHoliday> {
+    if (isEmpty(id)) throw new HttpException(400, 'provide Id');
+
+    const publicHoliday = await this.publicHolidayModel.findOne({ _id: id, deleted: false });
```
Comment (anthugny) [inline src/services/public-holiday/public-holiday.service.ts:53]: You can use findOneandUpdate query method.

File: src/services/public-holiday/public-holiday.service.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      throw new HttpException(404, `Public holiday not found`);
+    }
+    const updatedPublicHoliday = await publicHoliday.updateOne(updatedData);
+    return updatedPublicHoliday;
+  }
+
+  public async delete(id: string): Promise<IPublicHoliday> {
+    if (isEmpty(id)) throw new HttpException(400, 'provide Id');
+
+    const publicHoliday = await this.publicHolidayModel.findOne({ _id: id, deleted: false });
```
Comment (anthugny) [inline src/services/public-holiday/public-holiday.service.ts:65]: use [findOneAndDelete()](https://www.mongodb.com/docs/manual/reference/method/db.collection.findOneAndDelete/#mongodb-method-db.collection.findOneAndDelete)

File: src/dtos/public-holiday/public_holiday.dto.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  public end_date: string;
+
+  @IsString()
+  public title: string;
+
+  @IsString()
+  @IsOptional()
+  public project_id: string;
+
+  @IsString()
```
Comment (anthugny) [inline src/dtos/public-holiday/public_holiday.dto.ts:17]: remove the updatedby, createdby and deletedby from the DTO. Rather Interface will have these options.

File: src/controllers/public-holiday/public-holiday.controller.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      const publicHolidayEntries = await this.publicHolidayService.findAll();
+      res.status(200).json({ data: publicHolidayEntries });
+
+      console.table(
+        publicHolidayEntries.map(entry => ({
+          title: entry.title,
+          deleted: entry.deleted,
+        })),
+      );
+      console.log(`========  ${publicHolidayEntries.length} public holidays found  =======`);
```
Comment (anthugny) [inline src/controllers/public-holiday/public-holiday.controller.ts:30]: remove logs statements

File: src/controllers/public-holiday/public-holiday.controller.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      const publicHolidayEntries = await this.publicHolidayService.findAllActive();
+      res.status(200).json({ data: publicHolidayEntries });
+
+      console.table(
+        publicHolidayEntries.map(entry => ({
+          title: entry.title,
+          deleted: entry.deleted,
+        })),
+      );
+      console.log(`========  ${publicHolidayEntries.length} public holidays found  =======`);
```
Comment (anthugny) [inline src/controllers/public-holiday/public-holiday.controller.ts:48]: remove logs statements

File: src/controllers/public-holiday/public-holiday.controller.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      const id: string = req.params.id;
+      const data = await this.publicHolidayService.findById(id);
+      res.status(200).json({ data: data });
+
+      console.table({
+        title: data.title,
+        deleted: data.deleted,
+      });
+      console.log(`${data.title} public holidays found`);
+      console.log(`================================`);
```
Comment (anthugny) [inline src/controllers/public-holiday/public-holiday.controller.ts:66]: remove logs statements

**Review by anthugny — CHANGES_REQUESTED** [review]
Refer to the comments I left

File: src/services/public-holiday/public-holiday.service.ts (hunk 4 above)
Comment (clinton-otse) [reply src/services/public-holiday/public-holiday.service.ts:53]: image

**Pull request merged**